File: Source/Aladdin.ASN1.ISO/ASN1/ISO/OCSP/TBSRequest.cs
using System;
using System.Runtime.Serialization;

// TBSRequest ::= SEQUENCE {
//     version           [0] EXPLICIT INTEGER DEFAULT v1(0),
//     requestorName     [1] EXPLICIT GeneralName OPTIONAL,
//     requestList           SEQUENCE OF Request,
//     requestExtensions [2] EXPLICIT Extensions {{re-ocsp-nonce | re-ocsp-response, ..., re-ocsp-preferred-signature-algorithms}} OPTIONAL
//  }

namespace Aladdin.ASN1.ISO.OCSP
{
	[Serializable]
	public class TBSRequest : Sequence
	{
		// информация о структуре
		private static readonly ObjectInfo[] info = new ObjectInfo[] { 

			new ObjectInfo(new ObjectCreator<Integer          >().Factory(), Cast.E,  Tag.Context(0), new Integer(0)), 
			new ObjectInfo(new ChoiceCreator<PKIX.GeneralName >().Factory(), Cast.EO, Tag.Context(1)                ), 
			new ObjectInfo(new ObjectCreator<Sequence<Request>>().Factory(), Cast.N								    ), 
			new ObjectInfo(new ObjectCreator<PKIX.Extensions  >().Factory(), Cast.EO, Tag.Context(2)			    ) 
		}; 
		// конструктор при сериализации
        protected TBSRequest(SerializationInfo info, StreamingContext context) : base(info, context) {}

		// конструктор при раскодировании
		public TBSRequest(IEncodable encodable) : base(encodable, info) {} 

		// конструктор при закодировании
		public TBSRequest(Integer version, IEncodable requestorName, 
			Sequence<Request> requestList, PKIX.Extensions requestExtensions) : 
			base(info, version, requestorName, requestList, requestExtensions) {} 

		public Integer				Version				{ get { return (Integer				)this[0]; } } 
		public IEncodable			RequestorName		{ get { return (IEncodable			)this[1]; } }
		public Sequence<Request>	RequestList			{ get { return (Sequence<Request>	)this[2]; } }
		public PKIX.Extensions		RequestExtensions	{ get { return (PKIX.Extensions		)this[3]; } }
	}
}

File: Source/Aladdin.ASN1.ISO.Java/src/aladdin/asn1/iso/ocsp/TBSRequest.java
package aladdin.asn1.iso.ocsp;
import aladdin.asn1.*;
import aladdin.asn1.iso.pkix.*;
import java.io.*;

// TBSRequest ::= SEQUENCE {
//     version           [0] EXPLICIT INTEGER DEFAULT v1(0),
//     requestorName     [1] EXPLICIT GeneralName OPTIONAL,
//     requestList           SEQUENCE OF Request,
//     requestExtensions [2] EXPLICIT Extensions {{re-ocsp-nonce | re-ocsp-response, ..., re-ocsp-preferred-signature-algorithms}} OPTIONAL
//  }

public class TBSRequest extends Sequence<IEncodable>
{
    private static final long serialVersionUID = -1265760254928827535L;

    // информация о структуре
	private static final ObjectInfo[] info = new ObjectInfo[] { 

		new ObjectInfo(new ObjectCreator  (aladdin.asn1.Integer.class).factory(), Cast.E , Tag.context(0), new aladdin.asn1.Integer(0)), 
		new ObjectInfo(new ChoiceCreator  (GeneralName         .class).factory(), Cast.EO, Tag.context(1)                             ), 
		new ObjectInfo(new SequenceCreator(Request             .class).factory(), Cast.N                                              ), 
		new ObjectInfo(new ObjectCreator  (Extensions          .class).factory(), Cast.EO, Tag.context(2)                             ) 
	}; 
	// конструктор при раскодировании
	public TBSRequest(IEncodable encodable) throws IOException { super(encodable, info); }
    
	// конструктор при закодировании
	public TBSRequest(aladdin.asn1.Integer version, IEncodable requestorName, 
        Sequence<Request> requestList, Extensions requestExtensions) 
	{ 
		super(info, version, requestorName, requestList, requestExtensions); 
	}
	public final aladdin.asn1.Integer version           () { return (aladdin.asn1.Integer)get(0); } 
	public final IEncodable           requestorName     () { return                       get(1); }
    @SuppressWarnings({"unchecked"}) 
	public final Sequence<Request>    requestList       () { return (Sequence<Request>   )get(2); }
	public final Extensions           requestExtensions () { return (Extensions          )get(2); }
}
